feature/KAS-3405 Pub v3.1: Kunnen verwijderen van doc na ontvangst

mdv-professional opened the pull request:
[kasticketje](https://kanselarij.atlassian.net/browse/KAS-3405)

Commit:
KAS-3405 add deleteReceivedPiece /publications/:id/translations + /publications/:id/proofs
feature/KAS-3405-pub-v3.1--Verwijder-doc-na-ontvangst

Commit:
KAS-3405 add deleteReceivedPiece /publications/:id/translations + /publications/:id/proofs
feature/KAS-3405-pub-v3.1--Kunnen-verwijderen-van-doc-van-ontvangst

erikap commented on app/pods/publications/publication/proofs/index/controller.js:
```diff
@@ -88,6 +88,23 @@ export default class PublicationsPublicationProofsController extends Controller
+  @task
+  *deleteReceivedPiece(proofReceivedEvent, piece) {
+    yield this.performDeleteReceivedPiece(proofReceivedEvent, piece);
+    this.send('refresh');
```
Is deze refresh noodzakelijk? Er worden geen activities verwijderd, enkel mogelijks een end-date aangepast. Aangezien er gebruik gemaakt wordt van `get isShown()` in de `TimelineActivity` lijkt de view mij correct te updaten?

erikap commented on app/pods/publications/publication/proofs/index/controller.js:
```diff
@@ -88,6 +88,23 @@ export default class PublicationsPublicationProofsController extends Controller
+    await this.publicationService.deletePiece(piece);
+    const proofingActivity = proofReceivedEvent.activity;
+    let pieces = await proofingActivity.generatedPieces;
+    pieces = pieces.toArray();
+    if (pieces.length === 0) {
```
`toArray()` lijkt me niet noodzakelijk om `pieces.length` te kunnen doen?

erikap commented on app/pods/publications/publication/proofs/index/route.js:
```diff
@@ -6,9 +6,25 @@ import { warn } from '@ember/debug';
+
+  async initFields() {
+    if (this.isProofingActivity) {
+      let pieces = await this.activity.generatedPieces;
+      pieces = pieces.toArray();
+      let publicationActivities = await Promise.all(
+        pieces.map((piece) => piece.publicationActivitiesUsedBy)
+      );
+      publicationActivities = publicationActivities.flatMap((publicationActivities) =>
+        publicationActivities.toArray()
+      );
+      this.canDeletePieces = publicationActivities.length === 0;
```
Het is me uit het Jira ticket niet helemaal duidelijk of de delete-flag bepaald wordt per activity of per document. Ik zou per document logischer vinden, maar hier is precies gekozen voor activity. 
Als het op niveau van activity bepaald moet worden, is het waarschijnlijk efficientier (in aantal network requests) om hier een `queryOne` te doen met een filter op de piece ids en op het bestaan van de relatie voor `publicationActivitiesUsedBy`. Als er zo minstens 1 document bestaat, moet delete gedisabled worden voor de hele activiteit.

mdv-professional replied:
Ik weet ook niet wat hier verwacht wordt. Ik heb het aantal netwerk-requests voorlopig beperkt door deze relatie toe te voegen aan de include. Ik herinner me dat @MikiDi hier een voorkeur voor had in de rapporten-tab.

Ik besef dat per piece deleten soms nuttig kan zijn (vb. wanneer na de proofing-request nog een translation-document wordt geupload). Dit vergt wel wat refactoring. Mocht dit het geval zijn, is mijn voorstel om aan de `<PublicationDocumentsList />` een optioneel yield-block te kunnen meegeven om per piece attributes te kunnen meegeven i.p.v. voor heel de lijst.
vb.
```hbs
<PublicationDocumentsList @rows={{this.pieceRows}} as |pieceRow|>
  <PublicationDocumentItem @onDeletePiece={{if pieceRow.canDelete (fn this.deletePiece pieceRow)}} />
</PublicationDocumentsList>
```
Ik vind dit niet de meest vanzelfsprekende opzet van zo'n component, maar zie niet meteen een andere mogelijkheid. Om dit iets te versimpelen, en het gebruik van row-objecten te vermijden, ben ik nog steeds voorstander van meer gebruik van de `compute` helper in [ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers#compute), maar ik weet dat jullie hier geen voorstander van zijn.

tdn replied:
Ik denk dat we het hier ingewikkelder maken dan het moet zijn. Johan vat de nood van de gebruikers samen: "Het gaat om de fase vóór het verzenden. Ze hebben gewoon een verkeerd document van de harde schijf naar Kaleidos opgeladen en opgeslagen. Daar moet een vuilbakje komen zodat je dat opladen ongedaan kan maken."

Dus de situatie waarbij een document moet verwijderd worden dat al gebruikt is voor een vertaalaanvraag/drukproefaanvraag gaat zich eigenlijk nooit voordoen.

Wat wel kan is dat er nadat de vertaling/drukproef dan is ontvangen, er een verkeerd document wordt opgeslagen, dat moet dus wel kunnen verwijderd worden.

M.a.w., zolang je altijd documenten kan verwijderen in de panelen met als titel "Vertaling/drukproef ontvangen", is de vraag van de gebruikers voldaan. Is dat nu al zo?

erikap reviewed (commented):
Feedback gezet enkel op proof-related routes, maar analoge comments gelden voor translation-routes.

Commit:
KAS-3405 no refresh /publicaties/x/vertalingen & /publicaties/x/drukproeven
> Pub v3.1: Verwijder doc na ontvangst

Commit:
KAS-3405 no refresh /publicaties/x/vertalingen & /publicaties/x/drukproeven
> Pub v3.1: Verwijder doc na ontvangst

Commit:
KAS-3405 Timeline.create /publicaties/x/vertalingen & /publicaties/x/drukproeven
> Pub v3.1: Verwijder doc na ontvangst

ValenberghsSven reviewed (commented):
Het verwijderen van zo een document resulteert in het "verdwijnen" van de panel "vertaling ontvangen".
Zou het beter zijn om dan eerder een lege lijst te tonen of de gehele vertaling activity op te kuisen?
als je bvb vertaling upload, dan verwijdert, dan een aanvraag doet en een nieuwe vertaling upload, dan bestaat die eerste vertaling activity nog, zonder used-pieces.

erikap commented:
> Het verwijderen van zo een document resulteert in het "verdwijnen" van de panel "vertaling ontvangen". Zou het beter zijn om dan eerder een lege lijst te tonen of de gehele vertaling activity op te kuisen? als je bvb vertaling upload, dan verwijdert, dan een aanvraag doet en een nieuwe vertaling upload, dan bestaat die eerste vertaling activity nog, zonder used-pieces.

De vertaalactiviteit die steeds bestaat, is een model keuze die gemaakt is in het verleden. Aanvraag- en vertaalactiviteit komen steeds in paren voor. De vertaalactiviteit wordt al aangemaakt (zonder einddatum) zodra de aanvraag verstuurd wordt. Vertaalactiviteiten zonder einddatum werden voorheen ook al verborgen. Dat gedrag zou ik in deze PR niet omgooien.

ValenberghsSven commented:
> > Het verwijderen van zo een document resulteert in het "verdwijnen" van de panel "vertaling ontvangen". Zou het beter zijn om dan eerder een lege lijst te tonen of de gehele vertaling activity op te kuisen? als je bvb vertaling upload, dan verwijdert, dan een aanvraag doet en een nieuwe vertaling upload, dan bestaat die eerste vertaling activity nog, zonder used-pieces.
> 
> De vertaalactiviteit die steeds bestaat, is een model keuze die gemaakt is in het verleden. Aanvraag- en vertaalactiviteit komen steeds in paren voor. De vertaalactiviteit wordt al aangemaakt (zonder einddatum) zodra de aanvraag verstuurd wordt. Vertaalactiviteiten zonder einddatum werden voorheen ook al verborgen. Dat gedrag zou ik in deze PR niet omgooien.

Akkoord van het niet omgooien
Ik kan erop zich mee leven dat het blijft zoals het is, wou het gewoon even aanbrengen.

Commit:
Merge branch 'development' into feature/KAS-3405-pub-v3.1--Verwijder-doc-na-ontvangst

Commit:
Merge branch 'f/KAS-3405-pub-v3.1--Verwijder-doc-na-ontvangst--TimelineActivity.create' into feature/KAS-3405-pub-v3.1--Verwijder-doc-na-ontvangst

Commit:
KAS-3405 add x-activities-used-by to include /publicaties/x/vertalingen & /publicaties/x/drukproeven
> Pub v3.1: Verwijder doc na ontvangst

Commit:
KAS-3405 show empty message when only TimelineActivity is hidden
/publicaties/x/vertalingen & /publicaties/x/drukproeven
> Pub v3.1: Verwijder doc na ontvangst

ghost commented:
@ValenberghsSven Dit was eenvoudig te verhelpen: [zie commit](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1357/commits/74f1a4f1ad46b7cc2073363ceda3b0f6cb385c21)

tdn commented on app/pods/publications/publication/proofs/index/template.hbs:
```diff
@@ -38,41 +38,43 @@
+              @publicationFlow={{this.publicationFlow}}
+              @proofPrintCorrector={{this.publicationSubcase.proofPrintCorrector}}
+              @onEditProofingActivity={{perform this.editProofingActivity}}
+              @onDeletePiece={{if
+                event.canDeletePieces
+                (perform this.deleteReceivedPiece event)
+              }}
```
Vraagje hierbij (wellicht door mijn gebrek aan ervaring met ember): de delete-knop in `PublicationDocumentsList` wordt enkel getoond als `@onDeletePiece` gebonden/truthy is, toch? Zie https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/9e8f704ae15705b8d46ea157f567dbfe03a4ecd2/app/components/publications/publication/publication-documents-list.hbs#L42-L50

Als je `@onDeletePiece` hier op deze manier instelt, gaat hij dan de delete-knop ook tonen als `event.canDeletePieces` falsy is? Want dat is niet de bedoeling...

mdv-professional replied:
De delete-knop wordt niet getoond als `@onDeletePiece` false/undefined is. (Ik zou dit eigenlijk aanpassen, zodat de knop in dat geval gedisabled is. Dat lijkt me UX-gewijs correcter en meer in lijn met andere knoppen, maar da's een ander punt.)
